File: 剑指Offer/33-verify-post-order.cpp
#include <stack>
#include <iostream>
#include <vector>
#include <map>
#include <queue>
#include <utility>
#include <cstdint>
#include <algorithm>
#include <cstring>
using namespace std;


class Solution {
public:
    bool verifyPostorder(vector<int>& postorder) {

        int len = postorder.size();

        if(len<=0){
            return true;
        }

        return recursive(postorder,0,len-1);
    }

    bool recursive(vector<int>& postorder, int l, int r){

        if(l>=r)
            return true;

        int root = postorder[r];

        int i=l;
        for(;i<r;i++){  
            if(postorder[i]>root)
                break;
        }
        int p=i++;

        for(;i<r;i++){
            if(postorder[i]<root)
                return false;
        }

        return recursive(postorder,l,p-1) && recursive(postorder,p,r-1);
    }
};

File: 剑指Offer/33-verify-postorder.java
class Solution {
    public boolean verifyPostorder(int[] postorder) {

        
        int len = postorder.length;

        if(len<=0)
            return true;

        return verify(postorder, 0, len-1);

    }

    private boolean verify(int []postorder, int l, int r){
        if(l>=r){
            return true;
        }
        int root = postorder[r];
        int i = l;
        for(;i<r;i++){
            if(postorder[i]>root)
                break;
        }

        int p = i++;
        for(;i<r;i++)
            if(postorder[i]<root)
                return false;

        return verify(postorder,l,p-1) && verify(postorder,p,r-1);
    }
}